proof/courier.go

```diff
@@ -849,6 +921,206 @@ func (h *HashMailCourier) SetSubscribers(
 var _ Courier = (*HashMailCourier)(nil)
 
+// UniverseRpcCourier is a universe RPC proof courier service handle. It
+// implements the Courier interface.
+type UniverseRpcCourier struct {
+	// recipient describes the recipient of the proof.
+	recipient Recipient
+
+	// client is the RPC client that the courier will use to interact with
+	// the universe RPC server.
+	client unirpc.UniverseClient
+
+	// deliveryLog is the log that the courier will use to record the
+	// attempted delivery of proofs to the receiver.
+	deliveryLog DeliveryLog
+
+	// subscribers is a map of components that want to be notified on new
+	// events, keyed by their subscription ID.
+	subscribers map[uint64]*fn.EventReceiver[fn.Event]
+
+	// subscriberMtx guards the subscribers map and access to the
+	// subscriptionID.
+	subscriberMtx sync.Mutex
+}
+
+// DeliverProof attempts to delivery a proof file to the receiver.
+func (c *UniverseRpcCourier) DeliverProof(ctx context.Context,
+	annotatedProof *AnnotatedProof) error {
+
+	// Decode annotated proof into proof file.
+	proofFile := &File{}
+	err := proofFile.Decode(bytes.NewReader(annotatedProof.Blob))
+	if err != nil {
+		return err
+	}
+
+	// Iterate over each proof in the proof file and submit to the courier
+	// service.
+	for i := 0; i < proofFile.NumProofs(); i++ {
+		transitionProof, err := proofFile.ProofAt(uint32(i))
+		if err != nil {
+			return err
+		}
+		proofAsset := transitionProof.Asset
+
+		// Construct asset leaf.
+		rpcAsset, err := taprpc.MarshalAsset(
+			ctx, &proofAsset, true, true, nil,
+		)
+		if err != nil {
+			return err
+		}
+
+		var proofBuf bytes.Buffer
+		if err := transitionProof.Encode(&proofBuf); err != nil {
+			return fmt.Errorf("error encoding proof file: %w", err)
+		}
+
+		assetLeaf := unirpc.AssetLeaf{
+			Asset:         rpcAsset,
+			IssuanceProof: proofBuf.Bytes(),
+		}
+
+		// Construct universe key.
+		outPoint := transitionProof.OutPoint()
+		assetKey := unirpc.MarshalAssetKey(
+			outPoint, proofAsset.ScriptKey.PubKey,
+		)
+		universeID := unirpc.MarshalUniverseID(proofAsset.ID())
+		universeKey := unirpc.UniverseKey{
+			Id:      universeID,
+			LeafKey: assetKey,
+		}
+
+		// Submit proof to courier.
+		_, err = c.client.InsertProof(ctx, &unirpc.AssetProof{
+			Key:       &universeKey,
+			AssetLeaf: &assetLeaf,
+		})
+		if err != nil {
+			return fmt.Errorf("error inserting proof into "+
+				"universe courier service: %w", err)
+		}
+	}
+
+	return err
+}
+
+// ReceiveProof attempts to obtain a proof file from the courier service. The
+// final proof in the target proof file is identified by the given locator.
+func (c *UniverseRpcCourier) ReceiveProof(ctx context.Context,
+	originLocator Locator) (*AnnotatedProof, error) {
+
+	// Sleep for a bit to give the sender sufficient time to transfer the
+	// proof to the proof courier service.
+	//
+	// TODO(ffranr): Replace this with the backoff procedure as found in the
+	// hashmail courier.
+	time.Sleep(5 * time.Second)
```

Comment:
Why's this needed at all? Seems like an itest quirk? Why can't we rely on the normal back off stuff?

Reply:
The custodian waits for a relevant on-chain output and if found attempts to retrieve a proof file from the proof courier. But the sending node might still be in the middle of submitting a proof file one proof at a time to the proof courier service. Which means that the entire proof file (and the latest proof prefix) isn't available yet. So the receiving node waits before attempting to retrieve the proof file.

We need some sort of pause before trying to retrieve a proof file because (if all parties are involved are online) we need a time window for the sender to submit the proof and the courier to validate and serve. Otherwise the receiver's first attempt at retrieval will likely always fail. The backoff procedure does not include that delay but would overcome this issue by failing once and retrying.

We should have a constant for the `time.Sleep(5 * time.Second)` delay, I'll add that. I'm also going to move the sleep into the custodian.

I'll see what I can do about using the backoff procedure within the universe RPC courier, but it probably would be best to do that in a different PR.